bundle/jsky.app.ot/src/main/scala/jsky/app/ot/editor/seq/ItcTable.scala

```diff
@@ -55,29 +67,78 @@ trait ItcTable extends Table {
     ConfigBridge.extractSequence(_, null, ConfigValMapInstances.IDENTITY_MAP, true)
   }
 
+  protected def calculateSpectroscopy(peer: Peer, c: ItcUniqueConfig): Future[ItcService.Result] =
+    Future {
+      List("Not Implemented Yet").fail
+    }
+
+  protected def calculateImaging(peer: Peer, ins: SPComponentType, c: ItcUniqueConfig): Future[ItcService.Result] =
+    ins match {
+      case INSTRUMENT_GMOS | INSTRUMENT_GMOSSOUTH => calculateImagingGmos(peer, ins, c)
+      case _                                      => Future { List("Not Implemented Yet").fail }
+    }
+
+  protected def calculateImagingGmos(peer: Peer, ins: SPComponentType, c: ItcUniqueConfig): Future[ItcService.Result] = {
+    val port = owner.getContextIssPort
+    //val wfs  = ??? TODO
+    val src  = new SourceDefinition(PointSource(20.0, BrightnessUnit.MAG), LibraryStar("A0V"), WavebandDefinition.R, 0.0)
+    val obs  = new ObservationDetails(ImagingSN(c.count, c.singleExposureTime, 1.0), AutoAperture(5.0))
+    val tele = new TelescopeDetails(TelescopeDetails.Coating.SILVER, port, Wfs.OIWFS)
+
+    val qual = owner.getContextSiteQuality
+    val cond = new ObservingConditions(qual.getImageQuality, qual.getCloudCover, qual.getWaterVapor, qual.getSkyBackground, 1.5)
+
+    // get the instrument configuration
+    val filter    = c.config.getItemValue(new ItemKey("instrument:filter")).asInstanceOf[GmosCommonType.Filter]
```

Comment:
Sorry, I worded that text very poorly. My point was just that **if** you abstract out the `get` operation then you end up casting to a type variable, which is unsound, and you can use a `ClassTag` to get around this problem.

Here is an example of what I mean:

```scala
scala> def hmm[A](a: Any) = \/.fromTryCatch(a.asInstanceOf[A])
hmm: [A](a: Any)scalaz.\/[Throwable,A]

scala> hmm[Int]("hello")
res15: scalaz.\/[Throwable,Int] = \/-(hello) // <-- WAT .. this is mistyped

scala> res15.toOption
res16: Option[Int] = Some(hello) // <-- still mistyped

scala> res16.map(_ + 1) // finally an operation that catches us
java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Integer
  at scala.runtime.BoxesRunTime.unboxToInt(BoxesRunTime.java:105)
  at $anonfun$1.apply(<console>:18)
  at scala.Option.map(Option.scala:146)
  ... 43 elided
```

As opposed to

```scala
scala> def hmm[A](a: Any)(implicit ev: ClassTag[A]) = \/.fromTryCatch(ev.runtimeClass.cast(a).asInstanceOf[A])
hmm: [A](a: Any)(implicit ev: scala.reflect.ClassTag[A])scalaz.\/[Throwable,A]

scala> hmm[Int]("hello")
res19: scalaz.\/[Throwable,Int] = -\/(java.lang.ClassCastException: Cannot cast java.lang.String to int)
```